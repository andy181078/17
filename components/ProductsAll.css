.ProductsAll {
  border: 1px solid blueviolet;
  border-collapse: collapse;
  
}

.ProductsAll .Main {
  border: 1px solid blueviolet;
  color: green;
  padding: 3px;
  font-size: 20px;
  
}
