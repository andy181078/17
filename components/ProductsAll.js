import React from "react";
import PropTypes from 'prop-types';

import "./ProductsAll.css";
import {mobileEvents} from './events';

import ProductRow from "./ProductRow";
import ProductEdit from "./ProductEdit";
import ProductCard from "./ProductCard";

class ProductsAll extends React.PureComponent{
    static propTypes={
        goods: PropTypes.array.isRequired,
    }

    state ={
        goods: this.props.goods,
        selectedProductCode: null,
        cardMode: 0,  // 0 - ничего не выводим, 1 - просмотр карточки товара, 2 - редактирование и создание товара
        isValid: true, 
        isChanged: false, 
        isAdd: false, 
        isEdit: false, 
        isDelete: false,
    }

   

    componentDidMount =()=>{
        mobileEvents.addListener('EvEdit',this.evEdit);
        mobileEvents.addListener('EvDelete',this.evDelete);
        mobileEvents.addListener('EvCardView',this.evCardView);
        mobileEvents.addListener('EvChanged',this.evChanged);
        mobileEvents.addListener('EvSave',this.evSave);
        mobileEvents.addListener('EvAdd',this.evAdd);
        mobileEvents.addListener('EvCancel',this.evCancel);

        
        
    }

    componentWillUnmount =()=>{
        mobileEvents.removeListener('EvEdit',this.evEdit);
        mobileEvents.removeListener('EvDelete',this.evDelete);
        mobileEvents.removeListener('EvCardView',this.evCardView);
        mobileEvents.removeListener('EvChanged',this.evChanged);
        mobileEvents.removeListener('EvSave',this.evSave);
        mobileEvents.removeListener('EvAdd',this.evAdd);
        mobileEvents.removeListener('EvCancel',this.evCancel);
    }
    

    


    evChanged=(bool)=>{
        this.setState({isChanged: bool})
    }

    evCardView =(code)=>{
        if (!this.state.isChanged){
            this.setState({
                cardMode: 1, 
                selectedProductCode: code, 
                isEdit: false,
                
            })
        }
    }

    evEdit=(code)=>{
        if (!this.state.isChanged){
            this.setState({
                cardMode: 2, 
                selectedProductCode: code,
                isEdit: true
            })
        }
    }

   
    add=()=>{
        if (!this.state.isChanged && !this.state.isEdit){
            let code=this.state.goods.length+1
            
            this.setState({
                selectedProductCode: code,
                cardMode: 2, 
                isAdd: true,
            })
        }
    }

    evSave=(res)=>{
        let edit=[...this.state.goods]
        edit=edit.map(v=>(v.code==res.code)?res:v)
        this.setState({
            cardMode: 0, 
            goods: edit,
            isEdit: false
        })
    }

    evAdd=(res)=>{
        this.state.goods.push(res)
        
       
        this.setState({
            cardMode: 0, 
            
        })
    }

    evDelete=(code)=>{
        if (!this.state.isChanged && !this.state.isEdit){
            if(confirm('Вы действительно хотите удалить товар?')){
                let res=[...this.state.goods];
                res=res.filter(v=>(v.code!==code));
                this.setState({
                    cardMode: 0,
                    goods: res
                   
                })
            } 
        }
    }

    evCancel=()=>{
        this.setState({ cardMode: 0,
                        isEdit: false})
    }

    render(){
        console.log("product allrender")

// ----------------------- ТАБЛИЦА ТОВАРОВ -----------------------//   
        var goodsRows=this.state.goods.map( v=>
            <ProductRow key={v.code} 
                        row={v} 
                        code={v.code} 
                        selectedProductCode={this.state.selectedProductCode}
                        isChanged={this.state.isChanged}
                        isEdit={this.state.isEdit}
                       
            />
        );

// СТРОКА ДЛЯ РАБОТЫ
let selectedRow=[...this.state.goods]
    selectedRow=selectedRow.find((v, i)=>v.code==this.state.selectedProductCode)

        return (
        <div>
            

            
            <table className='ProductsAll'>
                <tbody>
                
                
                    <tr> 
                        <th className='Main'>Наименование</th>
                        <th className='Main'>Цена</th>
                        <th className='Main'>URL</th>
                        <th className='Main'>Количество</th>
                        <th className='Main'>Исправить</th>
                        <th className='Main'>Удалить</th>
                    </tr>
                    
                    {goodsRows}
        
                </tbody>
            </table>
           
            <input type='button' value='Новый товар' onClick={this.add} disabled={this.state.isChanged||this.state.isEdit}/>

 

{/*----------------------- ПРОСМОТР КАРТОЧКИ -----------------------*/}            
            {/* 
                (this.state.cardMode=="1") &&
                <ProductCard row={selectedRow} 
                />
           */ } 

{/*----------------------- СОЗДАНИЕ И РЕДАКТИРОВАНИЕ -----------------------*/}
            {
                (this.state.cardMode=="2") &&
                <ProductEdit  
                            code={this.state.selectedProductCode} 
                            row={selectedRow} 
                            
                            
                            
                />
            }
        </div>
        )
    }
};

export default ProductsAll;
    