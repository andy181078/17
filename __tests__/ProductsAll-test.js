"use strict"
import React from 'react';
import renderer from 'react-test-renderer';

import ProductsAll from "../components/ProductsAll"
                            //импортируем файл который будем тестировать


                  
let productsArr = [{
                    code: 1,
                    name: "Xiaomi Redmi 9C",
                    price: "1450",
                    url: "https://xistore.by/upload/resize/element/40056/082/8c4bed7595c8d031c1fa3140f267e255_482_482_50@x2.webp",
                    quant: 25
                },
                {
                    code: 2,
                    name: "Xiaomi Redmi 9T",
                    price: "1300",
                    url: "https://xistore.by/upload/resize/element/54071/35d/233362a5bb439d7f0563d6cdf494eb2a_482_482_50@x2.webp",
                    quant: 44
                },
                {
                    code: 3,
                    name: "Xiaomi Redmi Note 10S",
                    price: "1500",
                    url: "https://xistore.by/upload/resize/element/65048/aef/c525eee0d25db5d15c35bfadaabb8874_482_482_50@x2.webp",
                    quant: 10
                },
                {
                    code: 4,
                    name: "Xiaomi Redmi 9A",
                    price: "1550",
                    url: "https://xistore.by/upload/resize/element/25900/a59/1fa5c9ab7c694e4676d8fff64e7bcfc5_482_482_50@x2.webp",
                    quant: 10
                }
                
                
                ]

test("работа кнопки Add", ()=>{
                              // создаём тестовую версию компонента который будем тестировать
    const component = renderer.create(
        <ProductsAll  goods={productsArr}/> // с нужными данными для построения
    );

                            // 1. получаем снэпшот (HTML-снимок) компонента для сверки, чтобы вёрстка не испортилась
    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

                            // имитируем добавление нового клиента
    let newClient={code: 5, name: "vvv", price: "123", url: "Рkkk", quant: 400};
    component.getInstance().evAdd(newClient); //evAdd=(newRow)=> функция добавления 

                            //получаем снимок с новым клиентом
    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
})

test("работа кнопки Delete", ()=>{
  // создаём тестовую версию компонента который будем тестировать
    const component = renderer.create(
    <ProductsAll  goods={productsArr}/> // с нужными данными для построения
    );

    // 1. получаем снэпшот (HTML-снимок) компонента для сверки, чтобы вёрстка не испортилась
    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();

    // имитируем удаление клиента
    component.getInstance().evDelete(2); //evDelete=(code)=> функция удаления 

    //получаем снимок с новым клиентом
    componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
})



