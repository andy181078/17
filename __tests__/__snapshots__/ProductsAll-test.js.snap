// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа кнопки Add 1`] = `
<div>
  <table
    className="ProductsAll"
  >
    <tbody>
      <tr>
        <th
          className="Main"
        >
          Наименование
        </th>
        <th
          className="Main"
        >
          Цена
        </th>
        <th
          className="Main"
        >
          URL
        </th>
        <th
          className="Main"
        >
          Количество
        </th>
        <th
          className="Main"
        >
          Исправить
        </th>
        <th
          className="Main"
        >
          Удалить
        </th>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9C
        </td>
        <td
          className="Main"
        >
          1450
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/40056/082/8c4bed7595c8d031c1fa3140f267e255_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          25
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9T
        </td>
        <td
          className="Main"
        >
          1300
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/54071/35d/233362a5bb439d7f0563d6cdf494eb2a_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          44
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi Note 10S
        </td>
        <td
          className="Main"
        >
          1500
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/65048/aef/c525eee0d25db5d15c35bfadaabb8874_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9A
        </td>
        <td
          className="Main"
        >
          1550
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/25900/a59/1fa5c9ab7c694e4676d8fff64e7bcfc5_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Новый товар"
  />
</div>
`;

exports[`работа кнопки Add 2`] = `
<div>
  <table
    className="ProductsAll"
  >
    <tbody>
      <tr>
        <th
          className="Main"
        >
          Наименование
        </th>
        <th
          className="Main"
        >
          Цена
        </th>
        <th
          className="Main"
        >
          URL
        </th>
        <th
          className="Main"
        >
          Количество
        </th>
        <th
          className="Main"
        >
          Исправить
        </th>
        <th
          className="Main"
        >
          Удалить
        </th>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9C
        </td>
        <td
          className="Main"
        >
          1450
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/40056/082/8c4bed7595c8d031c1fa3140f267e255_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          25
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9T
        </td>
        <td
          className="Main"
        >
          1300
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/54071/35d/233362a5bb439d7f0563d6cdf494eb2a_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          44
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi Note 10S
        </td>
        <td
          className="Main"
        >
          1500
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/65048/aef/c525eee0d25db5d15c35bfadaabb8874_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9A
        </td>
        <td
          className="Main"
        >
          1550
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/25900/a59/1fa5c9ab7c694e4676d8fff64e7bcfc5_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Новый товар"
  />
</div>
`;

exports[`работа кнопки Delete 1`] = `
<div>
  <table
    className="ProductsAll"
  >
    <tbody>
      <tr>
        <th
          className="Main"
        >
          Наименование
        </th>
        <th
          className="Main"
        >
          Цена
        </th>
        <th
          className="Main"
        >
          URL
        </th>
        <th
          className="Main"
        >
          Количество
        </th>
        <th
          className="Main"
        >
          Исправить
        </th>
        <th
          className="Main"
        >
          Удалить
        </th>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9C
        </td>
        <td
          className="Main"
        >
          1450
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/40056/082/8c4bed7595c8d031c1fa3140f267e255_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          25
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9T
        </td>
        <td
          className="Main"
        >
          1300
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/54071/35d/233362a5bb439d7f0563d6cdf494eb2a_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          44
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi Note 10S
        </td>
        <td
          className="Main"
        >
          1500
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/65048/aef/c525eee0d25db5d15c35bfadaabb8874_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9A
        </td>
        <td
          className="Main"
        >
          1550
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/25900/a59/1fa5c9ab7c694e4676d8fff64e7bcfc5_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Новый товар"
  />
</div>
`;

exports[`работа кнопки Delete 2`] = `
<div>
  <table
    className="ProductsAll"
  >
    <tbody>
      <tr>
        <th
          className="Main"
        >
          Наименование
        </th>
        <th
          className="Main"
        >
          Цена
        </th>
        <th
          className="Main"
        >
          URL
        </th>
        <th
          className="Main"
        >
          Количество
        </th>
        <th
          className="Main"
        >
          Исправить
        </th>
        <th
          className="Main"
        >
          Удалить
        </th>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9C
        </td>
        <td
          className="Main"
        >
          1450
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/40056/082/8c4bed7595c8d031c1fa3140f267e255_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          25
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9T
        </td>
        <td
          className="Main"
        >
          1300
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/54071/35d/233362a5bb439d7f0563d6cdf494eb2a_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          44
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi Note 10S
        </td>
        <td
          className="Main"
        >
          1500
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/65048/aef/c525eee0d25db5d15c35bfadaabb8874_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr
        className="ProductRow"
        onClick={[Function]}
      >
        <td
          className="Main"
        >
          Xiaomi Redmi 9A
        </td>
        <td
          className="Main"
        >
          1550
        </td>
        <td
          className="Main"
        >
          https://xistore.by/upload/resize/element/25900/a59/1fa5c9ab7c694e4676d8fff64e7bcfc5_482_482_50@x2.webp
        </td>
        <td
          className="Main1"
        >
          10
        </td>
        <td
          className="Main"
        >
          <input
            className="EditButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Исправить"
          />
        </td>
        <td
          className="Main"
        >
          <input
            className="DelButton"
            disabled={false}
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Новый товар"
  />
</div>
`;
